### PR TITLE
deploy docs to server

### PR DESCRIPTION
**Pull request by trichter**
WIP

### CODE REVIEW

**Comment by ThomasLecocq** [comment]
just a note for later: we'll have to be careful with the ``no_ci`` tag and limit it to cases where the PR is **only** about documentation pages (e.g. changing the Contributing page etc), and be careful that there are no changes to the codebase which _could_ fail with regular tests.

**Comment by trichter** [comment]
> just a note for later: we'll have to be careful with the no_ci tag and limit it to cases where the PR is only about documentation pages (e.g. changing the Contributing page etc), and be careful that there are no changes to the codebase which could fail with regular tests.

True. I rather see it as a flag which can be used when work is still in progress. It should be removed at the end if any changes to the code base were introduced.

> And I would still run the lint_code part even with no_ci.

Good idea. Will do.

**Comment by d-chambers** [comment]
> True. I rather see it as a flag which can be used when work is still in progress. It should be removed at the end if any changes to the code base were introduced.

This is what I had in mind as well when creating the tag. If you are pushing a flurry of commits (https://xkcd.com/1296/) it seems nice to give the CI servers a break. Ideally, I would like to see the tests run before merging anything though.

**Comment by trichter** [comment]
This PR is ready for review and merging. Docs are deployed to docs.obspy.org. See the link in the checks tab for the doc for this PR. I also tested deploying to master directly from this branch (see docs.obspy.org/master). It appears to work as expected. Hopefully the release event can be tested soon, too...

**Comment by d-chambers** [comment]
Hey @trichter, at first glance it looks good to me but I am going to defer to either @ThomasLecocq  or @barsch for a proper review.

The docs build bot stuff runs server side no?

**Comment by trichter** [comment]
> The docs build bot stuff runs server side no?

Yes I started the `deploybot.py` script with nohup on the server.

**Comment by trichter** [comment]
I think it is better to post the status only when the bot finished its job. Will push and update.

File: misc/scripts/docs_deploybot/remove_old_pr_docs.py
Comment (ThomasLecocq) [inline misc/scripts/docs_deploybot/remove_old_pr_docs.py:11]: this should be removed?

File: misc/scripts/docs_deploybot/deploy_docs.py
Comment (ThomasLecocq) [inline misc/scripts/docs_deploybot/deploy_docs.py:24]: is it normal that the URL is not built from the PATH? (or only for pr) ?

**Review by ThomasLecocq — APPROVED** [review]

**Comment by ThomasLecocq** [comment]
is it possible to make the "deploy link" use target="_blank" ? :-)

File: misc/scripts/docs_deploybot/deploy_docs.py
Comment (trichter) [reply misc/scripts/docs_deploybot/deploy_docs.py:24]: Yes that is normal, it can be configured. The advantage is that the stable branch sits in an extra dictionary and not in root.

**Comment by trichter** [comment]
> is it possible to make the "deploy link" use target="_blank" ? :-)

It is possible, but do we really want this? All other links in this tab do not use target="_blank".

**Comment by ThomasLecocq** [comment]
> > is it possible to make the "deploy link" use target="_blank" ? :-)
> 
> It is possible, but do we really want this? All other links in this tab do not use target="_blank".

I wish they did 🤣

**Comment by ThomasLecocq** [comment]
Once the last leftover from your rebase is gone, it's good to go for me ! Amazing job on this @trichter , the doc-build-bot is back !!! ❤️ 


Todo next (when this is merged, I'd say, what do you think?): add the warnings.log with the built PR deployed, and write some clever parser to make the output readable for the authors, starting by cleaning the non-package warnings:

```python
#!/opt/miniconda3/envs/py38obs/bin/python
import sys

lines = open(sys.argv[-1], 'r').read().splitlines()

for line in lines:
    if line.count("docstring of textwrap") or line.count("docstring of collections") or line.count("packages/obspy.io.nied.knet' that doesn't have a title"):
        continue
    print(line)

```

**Comment by trichter** [comment]
I think everything should be working now. Lets see if everything turns green.

**Comment by trichter** [comment]
> Todo next (when this is merged, I'd say, what do you think?): add the warnings.log with the built PR deployed, and write some clever parser to make the output readable for the authors, starting by cleaning the non-package warnings

Nice to have, but we can see all doc build warnings in github actions, cant we?

**Comment by ThomasLecocq** [comment]
> > Todo next (when this is merged, I'd say, what do you think?): add the warnings.log with the built PR deployed, and write some clever parser to make the output readable for the authors, starting by cleaning the non-package warnings
> 
> Nice to have, but we can see all doc build warnings in github actions, cant we?

No, it's really difficult to separate them from the logs, and I'd love to avoid reaching 3250 warnings again in the future... :-)

**Comment by trichter** [comment]
> No, it's really difficult to separate them from the logs, and I'd love to avoid reaching 3250 warnings again in the future... :-)

I see, but that is easy to have. Write all warnings to a file wit `-w`. You posted the command yourself somewhere. After that filter and print the warnings in a new step.

```
cat warnings.txt | grep obspy > obspy_warnings.txt
cat obspy_warnings.txt
echo `wc -l obspy_warnings.txt` warnings
```

Also, I think most warnings should also occur when running the tests and are reported to tests.obspy.org.

**Comment by ThomasLecocq** [comment]
> > No, it's really difficult to separate them from the logs, and I'd love to avoid reaching 3250 warnings again in the future... :-)
> 
> I see, but that is easy to have. Write all warnings to a file with `-w`. You posted the command yourself somewhere. After that filter and print the warnings in a new step.
> 
> ```
> cat warnings.txt | grep obspy > obspy_warnings.txt
> cat obspy_warnings.txt
> echo `wc -l obspy_warnings.txt` warnings
> ```
> 
> Also, I think most warnings should also occur when running the tests and are reported to tests.obspy.org.

The warning.log contains a lot of lines that don't have "obspy" in them (tracebacks from failed plots etc). So the full file is OK for uploading to the server, it'll allow parsing it later. The warning log contains also sphinx-related errors (missing index, wrong :type: etc), that are not tested using regultar tests, so it's good to have when someone pushes new code/doc.

Side note: I've started testing ``darglint`` as a very agressive linter for docstrings, it's quite good actually :-) I'll give it a pass on all the code base over the next weeks.

**Pull request merged**
